src/python/pants/engine/internals/rule_visitor.py

```diff
@@ -54,18 +56,72 @@ def _node_str(node: Any) -> str:
     return str(node)
 
 
+PANTS_RULE_DESCRIPTORS_MODULE_KEY = "__pants_rule_descriptors__"
+
+
+@dataclass(frozen=True)
+class RuleDescriptor:
+    """The data we glean about a rule by examining its AST.
+
+    This will be lazily invoked in the first `@rule` decorator in a module. Therefore it will parse
+    the AST *before* the module code is fully evaluated, and so the return type may not yet exist as
+    a parsed type. So we store it here as a str and look it up later.
+    """
+
+    module_name: str
+    rule_name: str
+    return_type: str
+
+    @property
+    def rule_id(self) -> str:
+        # TODO: Handle canonical_name/canonical_name_suffix?
+        return f"{self.module_name}.{self.rule_name}"
+
+
+def get_module_scope_rules(module: ModuleType) -> tuple[RuleDescriptor, ...]:
+    """Get descriptors for @rules defined at the top level of the given module.
+
+    We discover these top-level rules and rule helpers in the module by examining the AST.
+    This means that while executing the `@rule` decorator of a rule1(), the descriptor of a rule2()
+    defined later in the module is already known.  This allows rule1() and rule2() to be
+    mutually recursive.
+
+    TODO: Support mutual recursion for rules defined in inner scopes?
+    """
+    descriptors = getattr(module, PANTS_RULE_DESCRIPTORS_MODULE_KEY, None)
+    if descriptors is None:
+        descriptors = []
+        for node in ast.iter_child_nodes(ast.parse(inspect.getsource(module))):
+            if isinstance(node, ast.AsyncFunctionDef) and isinstance(node.returns, ast.Name):
+                descriptors.append(RuleDescriptor(module.__name__, node.name, node.returns.id))
+        descriptors = tuple(descriptors)
+        setattr(module, PANTS_RULE_DESCRIPTORS_MODULE_KEY, descriptors)
+
+    return descriptors
+
+
 class _TypeStack:
+    """The types and rules that a @rule can refer to in its input/outputs, or its awaitables.
+
+    We construct this data through a mix of inspection of types already parsed by Python,
+    and descriptors we infer from the AST. This allows us to support mutual recursion between
+    rules defined in the same module (the @rule descriptor of the earlier rule can know enough
+    about the later rule it calls to set up its own awaitables correctly).
+
+    This logic is necessarily heuristic. It works for well-behaved code, but may be defeated
```

Comment:
How would a failing heuristic be conveyed to the user? Would it just be the original circular ref error, or is there something more precise we could action them to?

Reply:
Same as before - circular ref error.